test/blocklib/core/selector.hpp

```diff
@@ -0,0 +1,158 @@
+
+// optional shortening
+template<typename T, fair::meta::fixed_string description = "", typename... Arguments>
+using A           = Annotated<T, description, Arguments...>;
+
+using SelectorDoc = Doc<R""(
+@brief basic multiplexing class to route arbitrary inputs to outputs
+)"">;
+
+template<typename T>
+struct Selector : node<Selector<T>, SelectorDoc> {
+    // port definitions
+    PortIn<std::uint32_t, Async, Optional> selectOut;
+    PortOut<T, Async, Optional>            monitorOut; // optional monitor output (more for demo/API purposes than actual need)
+    std::vector<PortIn<T, Async>>          inputs;     // TODO: need to add exception to pmt_t that this isn't interpreted as a settings type
+    std::vector<PortOut<T, Async>>         outputs;
+
+    // settings
+    A<std::uint32_t, "nInputs", Visible, Doc<"variable number of inputs">, Limits<1U, 32U>>    nInputs  = 0U;
+    A<std::uint32_t, "nOutputs", Visible, Doc<"variable number of inputs">, Limits<1U, 32U>>   nOutputs = 0U;
+    A<std::vector<std::uint32_t>, "mapIn", Visible, Doc<"input port index to route from">>     mapIn; // N.B. need two vectors since pmt_t doesn't support pairs (yet!?!)
+    A<std::vector<std::uint32_t>, "mapOut", Visible, Doc<"output port index to route to">>     mapOut;
+    A<bool, "backPressure", Visible, Doc<"true: do not consume samples from un-routed ports">> backPressure = false;
+    std::map<std::uint32_t, std::vector<std::uint32_t>>                                        _internalMapping;
+    std::uint32_t                                                                              _selectedSrc = -1U;
+
+    void
+    settings_changed(const fair::graph::property_map &old_settings, const fair::graph::property_map &new_settings) {
+        if (new_settings.contains("nInputs") || new_settings.contains("nOutputs")) {
+            fmt::print("{}: configuration changed: nInputs {} -> {}, nOutputs {} -> {}\n", static_cast<void *>(this), old_settings.at("nInputs"),
+                       new_settings.contains("nInputs") ? new_settings.at("nInputs") : "same", old_settings.at("nOutputs"), new_settings.contains("nOutputs") ? new_settings.at("nOutputs") : "same");
+            inputs.resize(nInputs);
+            outputs.resize(nOutputs);
+        }
+        if (new_settings.contains("mapIn") || new_settings.contains("mapOut")) {
+            assert(mapIn.value.size() == mapOut.value.size() && "mapIn and mapOut must have the same length");
+            _internalMapping.clear();
+
+            if (mapIn.value.size() != mapOut.value.size()) {
+                throw std::invalid_argument("Input and output map need to have the same number of elements");
+            }
+
+            for (std::size_t i = 0U; i < mapOut.value.size(); ++i) {
+                _internalMapping[mapIn.value[i]].push_back(mapOut.value[i]);
+            }
+        }
+    }
+
+    using select_reader_t  = typename PortIn<std::uint32_t, Async, Optional>::ReaderType;
```

Comment:
:thinking: I liked the concepts `ConsumableSpan auto` and `ProducableSpan auto&` better because the`ReaderType` and `WriterType` are `circular_buffer<T>` implementation specific and could be replaced by the user....

OK for now, will experiment with it separately.